crates/datasources/src/mongodb/mod.rs

```diff
@@ -271,18 +279,136 @@ impl TableProvider for MongoTableProvider {
         // Note that this projection will only project top-level fields. There
         // is not a way to project nested documents (at least when modelling
         // nested docs as a struct).
-        let projected_schema = match projection {
+        let schema = match projection {
             Some(projection) => Arc::new(self.schema.project(projection)?),
             None => self.schema.clone(),
         };
 
-        // TODO: Filters.
+        // Projection document. Project everything that's in the schema.
+        //
+        // The `_id` field is special and needs to be manually suppressed if not
+        // included in the schema.
+        let mut proj_doc = Document::new();
+        let mut has_id_field = false;
+        for field in &schema.fields {
+            proj_doc.insert(field.name(), 1);
+            has_id_field = has_id_field || field.name().as_str() == ID_FIELD_NAME;
+        }
+
+        if !has_id_field {
+            proj_doc.insert(ID_FIELD_NAME, 0);
+        }
+
+        let mut find_opts = FindOptions::default();
+        find_opts.limit = limit.map(|v| v as i64);
+        find_opts.projection = Some(proj_doc);
 
-        Ok(Arc::new(MongoBsonExec::new(
-            projected_schema,
-            self.collection.clone(),
-            limit,
-        )))
+        let filter = exprs_to_mdb_query(_filters).expect("could not build query");
+        let cursor = Arc::new(
+            self.collection
+                .find(Some(filter), Some(find_opts))?
+                .into()
+                .as_ref(),
+        );
+
+        Ok(Arc::new(MongoBsonExec::new(cursor, limit)))
```

Comment:
I don't think we should be creating a cursor in the `scan`. Instead we should push the filter down to the physical exec. That should also keep the Exec threadsafe.

Reply:
This doesn't seem to be consistent with how the code works, consistently: mysql and postgres operate the way you expect; bigquery and snowflake operate (roughly, to my reading) as here. 

It seems from what Sean said above not only does Exec not need to be thread safe, but that those methods aren't ever called more than once.